### PR TITLE
[Transform Dialect][LLVMGPU] Adding support for MatmulTensorCoreStrategy without linalg.fill

### PR DESCRIPTION
**Pull request by tavakkoliamirmohammad**
This patch enables tensor core strategy to target cases where `linalg.matmul` output is not allocated as a part of code using `linalg.fill` op. It will allow targeting broader cases of `linalg.matmul`.

### CODE REVIEW

**Comment by Sameeranjoshi** [comment]
This is the current error we are facing right now, after the current patch.
```
[u1418973@notch369:~]$ cat ${IREE_SAMPLES_DIR}/transform_dialect/examples/matmul.mlir |   sed "s/\${M}/510/g" | sed "s/\${N}/510/g" | sed "s/\${K}/510/g" |   sed "s/private @matmul_static(/@matmul_static(/g" |   ${LLVM_BUILD_DIR}/bin/mlir-opt -symbol-dce |   ${IREE_DIR}/tools/iree-opt      --iree-hal-target-backends=cuda    --iree-hal-cuda-llvm-target-arch=sm_80  --iree-abi-transformation-pipeline      --iree-flow-transformation-pipeline      --iree-stream-transformation-pipeline      --iree-hal-configuration-pipeline | ${IREE_DIR}/tools/iree-opt  --pass-pipeline="builtin.module(hal.executable(hal.executable.variant(iree-llvmgpu-lower-executable-target)))" --iree-codegen-llvmgpu-enable-transform-dialect-matmul-tensorcore-strategy --iree-codegen-llvmgpu-enable-transform-dialect-aligned-matmul
<stdin>:13:9: error: transform.iree.pipeline_shared_memory_copies failed to apply
        func.func @matmul_static_dispatch_0_matmul_510x510x510_f32() {
        ^
<stdin>:21:16: note: when applied to this op
          %6 = linalg.matmul ins(%3, %4 : tensor<510x510xf32>, tensor<510x510xf32>) outs(%5 : tensor<510x510xf32>) -> tensor<510x510xf32>
               ^

```

**Comment by Sameeranjoshi** [comment]
Repro commands for the facing issue.

1. Below command gives input file - **`matmul_no_fill.mlir`** note the no-fill.
```
cat ${IREE_SAMPLES_DIR}/transform_dialect/examples/matmul.mlir |   sed "s/\${M}/510/g" | sed "s/\${N}/510/g" | sed "s/\${K}/510/g" |   sed "s/private @matmul_static(/@matmul_static(/g" |   ${LLVM_BUILD_DIR}/bin/mlir-opt -symbol-dce &> matmul_no_fill.mlir
```

2. Below command gives Transform Dialect script + HAL 
```
cat ${IREE_SAMPLES_DIR}/transform_dialect/examples/matmul.mlir |   sed "s/\${M}/510/g" | sed "s/\${N}/510/g" | sed "s/\${K}/510/g" |   sed "s/private @matmul_static(/@matmul_static(/g" |   ${LLVM_BUILD_DIR}/bin/mlir-opt -symbol-dce |   ${IREE_DIR}/tools/iree-opt      --iree-hal-target-backends=cuda    --iree-hal-cuda-llvm-target-arch=sm_80  --iree-abi-transformation-pipeline      --iree-flow-transformation-pipeline      --iree-stream-transformation-pipeline      --iree-hal-configuration-pipeline | ${IREE_DIR}/tools/iree-opt  --pass-pipeline="builtin.module(hal.executable(hal.executable.variant(iree-llvmgpu-lower-executable-target{test-lowering-configuration})))" --iree-codegen-llvmgpu-enable-transform-dialect-matmul-tensorcore-strategy --iree-codegen-llvmgpu-enable-transform-dialect-aligned-matmul
```

3. I remove only  Transform dialect part from the above output into a file called `td.mlir`.

4. Run through iree-compile.
```
${IREE_DIR}/tools/iree-compile  **matmul_no_fill.mlir**            --iree-hal-target-backends=cuda --iree-hal-cuda-llvm-target-arch=sm_80         --iree-codegen-llvmgpu-enable-transform-dialect-jit=false            --iree-codegen-llvmgpu-use-transform-dialect=**td.mlir** &>delete.mlir
```
This gives the shared mem error as described in the above error in this patch.

To solve this I comment the line (line 172) `transform.iree.pipeline_shared_memory_copies`, as suggested in error message.
Ok so this raises the error as below 
```
<unknown>:0: error: LLVM Translation failed for operation: builtin.unrealized_conversion_cast
<unknown>:0: note: see current operation: %99 = "builtin.unrealized_conversion_cast"(%98) : (!llvm.struct<(ptr<3>, ptr<3>, i64, array<2 x i64>, array<2 x i64>)>) -> memref<128x128xf32, #gpu.address_space<workgroup>>
matmul_no_fill.mlir:3:10: error: failed to translate the MLIR LLVM dialect to the native llvm::Module
    %0 = linalg.matmul ins(%arg0, %arg1 : tensor<510x510xf32>, tensor<510x510xf32>) outs(%arg2 : tensor<510x510xf32>) -> tensor<510x510xf32>
         ^
matmul_no_fill.mlir:2:3: note: called from
  func.func @matmul_static(%arg0: tensor<510x510xf32>, %arg1: tensor<510x510xf32>, %arg2: tensor<510x510xf32>) -> tensor<510x510xf32> {
  ^
```

**Observation** - If I change the input from sed  's/private @matmul_static(/@matmul_static(/g'  to 'sed "s/private @fill_matmul_static(/@fill_matmul_static(/g"'  and follow steps it works and generates executable.

```
cat ${IREE_SAMPLES_DIR}/transform_dialect/examples/matmul.mlir |   sed "s/\${M}/510/g" | sed "s/\${N}/510/g" | sed "s/\${K}/510/g" |   sed "s/private @fill_matmul_static(/@fill_matmul_static(/g" |   ${LLVM_BUILD_DIR}/bin/mlir-opt -symbol-dce &> matmulfill.mlir
```

**Comment by Sameeranjoshi** [comment]
@MaheshRavishankar I don't have permissions to add reviewers can you please add @nicolasvasilache in this ?
Thanks

**Comment by nicolasvasilache** [comment]
Could you try rebasing on top of #14138 ?

Pipelining is quite brittle and will often fail to apply but this is fine, it just means the optimization will not be applied.

**Comment by tavakkoliamirmohammad** [comment]
@nicolasvasilache We rebased our patch on top of upstream IREE. We used the same commands https://github.com/openxla/iree/pull/14103#issuecomment-1592375857, the shared memory copy pipeline error is now gone but the unrealized cast error that prevents `iree-compile` generating binary still persists.

File: llvm-external-projects/iree-dialects/lib/Transforms/TransformMatchers.cpp [hunk 1]
```diff
@@ -1371,6 +1371,7 @@ void transform_ext::makeMatmulMatcher(
     transform_ext::MatcherContext &matcherContext,
     transform_ext::StructuredOpMatcher *&matmulCapture,
     transform_ext::StructuredOpMatcher *&fillCapture,
+    transform_ext::StructuredOpMatcher *&leadingCapture,
```
Comment (qedawkins) [inline llvm-external-projects/iree-dialects/lib/Transforms/TransformMatchers.cpp:1374]: Is the added leading capture on the LHS related to the support for pre-filled outputs?

**Review by qedawkins — CHANGES_REQUESTED** [review]
Thanks! There will need to be a test added for a matmul without the fill to:
https://github.com/openxla/iree/blob/main/compiler/src/iree/compiler/Codegen/LLVMGPU/test/set_transform_strategy.mlir

**Comment by benvanik** [comment]
Closing stale PR.

**Pull request closed without merging**